Env map importance sampling

kevyuu opened the pull request:
## Description
Rework environment map importance sampling to vulkan and hlsl

## Testing 
Rework example 0 to use vulkan and hlsl
[Unit Test Pull Request](https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/pull/257)

## TODO list:
- Implement Warpmap Generation in hlsl
- Implement Warpmap hierarchical map sampling in hlsl

Commit:
Fix small bug in GenericDataAccessor definition

Commit:
First draft of Warpmap Generation workgroup implementation

Commit:
Add warp concept

Commit:
Add spherical warp

Commit:
Remove envmap accessors.hlsl

Commit:
Hierarchical image sampling implementation

Commit:
Merge branch 'master' into env_map_importance_sampling

Commit:
Merge branch 'master' into env_map_importance_sampling

Commit:
Remove envmap.hlsl

Commit:
Move to sampling namespace and implement backward density

Commit:
Remove private, public from hierarchical_image

Commit:
Refactor hierarchical image to keep accessor and common data as member

Commit:
Refactor hierarchical image to separate binarySearch from HierarchicalImage class

Commit:
Fix Spherical warp indentation

Commit:
Add some comment why we add xi to the sample uvs

Commit:
Merge branch 'master' into env_map_importance_sampling

Commit:
WIP

Commit:
Rename uv to coord for LuminanceAccessor concepts

Commit:
Fix hierarchical_image.hlsl

Commit:
Fix typo in spherical.hlsl

Commit:
Implement gen_luma, gen_warpmap and measure_luma shaders

Commit:
EnvmapImportanceSampling CMakeLists

Commit:
Initial implementation of CEnvmapImportanceSampling

Commit:
Initial implementation of CEnvmapImportanceSampling

Commit:
Small fixes

Commit:
Initial implementation of computeWarpMap

Commit:
Fix arithmetic config no const specifier for method

Commit:
Define config_t from outside

Commit:
More fixes on computeWarpMap implementation

Commit:
Fix chose second to be placed inside the loop

Commit:
LuminanceReadAccessor take ScalarT as template parameter

Commit:
gen_warpmap to gen_warp

devshgraphicsprogramming commented on include/nbl/ext/EnvmapImportanceSampling/builtin/hlsl/common.hlsl:
```diff
@@ -0,0 +1,49 @@
+
+struct device_capabilities
+{
+#ifdef TEST_NATIVE
```
why are you using macros? Macros are banned !

kevyuu commented on include/nbl/ext/EnvmapImportanceSampling/builtin/hlsl/common.hlsl:
```diff
@@ -0,0 +1,49 @@
+
+struct SLumaMeasurePushConstants
+{
+  float32_t4 luminanceScales;
```
Done.

Commit:
Move hierarchical_image concepts to sampling subdirectory

Commit:
Add some comment regarding corner sampling

Commit:
Parameterize spherical warp and make sure all literal is in the correct type

Commit:
Refactor CEnvmapImportanceSampling to block and calculate avgLuma

keptsecret committed:
merge master, fix conflicts

Commit:
Fix warp concept and add density type to warp concept

Commit:
Rename luminanceScale to lumaRGBCoefficients

Commit:
Remove measure_luma.comp.hlsl

Commit:
Fix some bug in hierarchical_image.hlsl

Commit:
Rename luminanceScales to lumaRGBCoefficients

kevyuu commented on include/nbl/builtin/hlsl/sampling/warp.hlsl:
```diff
@@ -36,18 +35,14 @@ NBL_CONCEPT_BEGIN(3)
-    ((NBL_CONCEPT_REQ_TYPE)(U::codomain_type))
-    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((warper.template warp(xi)) , ::nbl::hlsl::is_same_v, WarpResult<typename U::codomain_type>))
-    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((warper.template forwardDensity(xi)) , ::nbl::hlsl::is_same_v, float32_t))
-    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((warper.template backwardDensity(dst)) , ::nbl::hlsl::is_same_v, float32_t))
```
Done.

Commit:
Move EnvmapImportanceSampling from ext to core

Commit:
Fix binarySearch implementation. when last is 2x1 we should check for 0,0 and 1,0 not 0,1

Commit:
Rename private member with underscore prefix

Commit:
Merge branch 'master' into env_map_importance_sampling

Commit:
Update submodule to follow master branch

Commit:
Add missed EnvmapSampler.h and cpp

Commit:
Rename get and gather to texelFetch and texelGather

Commit:
Merge branch 'hlsl_path_tracer_example' into env_map_importance_sampling

Commit:
Include missing files into commit

Commit:
Merge branch 'master' into env_map_importance_sampling

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/hierarchical_image.hlsl:
```diff
@@ -0,0 +1,193 @@
+	  return result;
+	}
+
+	static bool choseSecond(scalar_type first, scalar_type second, NBL_REF_ARG(scalar_type) xi)
```
prefix with `__` for private methods

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/hierarchical_image/accessors.hlsl:
```diff
@@ -0,0 +1,62 @@
+#undef level
+#undef coord
+#undef a
+#include <nbl/builtin/hlsl/concepts/__end.hlsl>
```
if you merge latest path tracer branch, you can forego this concept in lieu of a LoadableImage with Components=1 and  add a GatherableImage with all the other accessors , see https://github.com/Devsh-Graphics-Programming/Nabla/pull/969#discussion_r2792476834

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/hierarchical_image/accessors.hlsl:
```diff
@@ -0,0 +1,61 @@
+);
+#undef sampler
+#undef coord
+#include <nbl/builtin/hlsl/concepts/__end.hlsl>
```
cool now we have docs, but I think the name is wrong, its `HierarchicalSampler` but you only use it for a warpmap so should be in `hlsl/sampling/warp`

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/hierarchical_image/accessors.hlsl:
```diff
@@ -0,0 +1,62 @@
+#undef a
+#include <nbl/builtin/hlsl/concepts/__end.hlsl>
+
+// sampleUvs return 4 UVs in a square to calculate the jacobian matrix
```
replace "to calculate the jacobian matrix" with "for manual bilinear interpolation with differentiability"

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/hierarchical_image/gen_luma.comp.hlsl:
```diff
@@ -0,0 +1,25 @@
+[[vk::push_constant]] SLumaGenPushConstants pc;
+
+[[vk::binding(0, 0)]] Texture2D<float32_t4> envMap;
+[[vk::binding(1, 0)]] RWTexture2D<float32_t> outImage;
```
this assumes a 2D map and not a cubemap/layered thing, use layered images instead

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/hierarchical_image/gen_warp.comp.hlsl:
```diff
@@ -0,0 +1,48 @@
+            lumaMap.Load(uint32_t3(coord, level), uint32_t2(1, 1)),
+            lumaMap.Load(uint32_t3(coord, level), uint32_t2(1, 0)),
+            lumaMap.Load(uint32_t3(coord, level), uint32_t2(0, 0))
+        );
```
btw OOB reads using texelFetch are undefined, asser that `coord` is between 0 and lastPixel-1 for both axes

kevyuu replied:
Done.

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/hierarchical_image/accessors.hlsl:
```diff
@@ -0,0 +1,62 @@
+#define NBL_CONCEPT_TPLT_PRM_KINDS (typename)(typename)
+#define NBL_CONCEPT_TPLT_PRM_NAMES (HierarchicalSamplerT)(ScalarT)
+// not the greatest syntax but works
+#define NBL_CONCEPT_PARAM_0 (sampler,HierarchicalSamplerT)
```
sampler is a reserved keyword in HLSL

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/hierarchical_image.hlsl:
```diff
@@ -0,0 +1,193 @@
+		// We use _lastWarpPixel here for corner sampling
+    float32_t2 xi = float32_t2(coord)/ _lastWarpPixel;
+		uint32_t2 p = uint32_t2(0, 0);
+		const uint32_t2 mip2x1 = findMSB(_mapSize.y);
```
you also wouldn't store `_lastWarpPixel` here then, because it would just be a warpmap agnostic sampler

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/warp.hlsl:
```diff
@@ -0,0 +1,55 @@
+}
+}
+
+#endif
```
I think this file will be made obsolete by #1001 s Bijective Sampler  concept

kevyuu replied:
This is warp concept not sampling concept.

kevyuu replied:
Done. Removed.

devshgraphicsprogramming commented on include/nbl/core/sampling/EnvmapSampler.h:
```diff
@@ -0,0 +1,148 @@
+#include "nbl/video/declarations.h"
+
+namespace nbl::core
+{
```
completely wrong namespace, you're using video, this file needs to live in `nbl/video/sampling`

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/hierarchical_image.hlsl:
```diff
@@ -0,0 +1,193 @@
+
+		matrix<scalar_type, 4, 2> uvs = _warpMap.sampleUvs(uint32_t2(texelCoord));
+
+		const vector2_type interpolant = frac(texelCoord);
```
make the `_warpMap.gather` spit out the interpolant so its cleaner (requires a specialized gather overload, adding on top of Gatherable concept)

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/hierarchical_image.hlsl:
```diff
@@ -0,0 +1,193 @@
+	LuminanceAccessorT _lumaMap;
+	HierarchicalSamplerT _warpMap;
+	uint32_t2 _warpSize;
+	uint32_t2 _lastWarpPixel;
```
you actually only need to store the `_lastWarpPixel.x * _lastWarpPixel.y` prodct, I expect the warpmap to handle gathering

kevyuu replied:
Done.

Commit:
Update comment on sampleUvs

Commit:
Use bitfield for lumaMapResolution in push constant

Commit:
Remove NBL_BUILD_ENVMAP_IMPORTANCE_SAMPLING option

Commit:
Fix worgroup dim for gen_warp

kevyuu commented on include/nbl/builtin/hlsl/sampling/hierarchical_image/gen_luma.comp.hlsl:
```diff
@@ -0,0 +1,25 @@
+[[vk::binding(0, 0)]] Texture2D<float32_t4> envMap;
+[[vk::binding(1, 0)]] RWTexture2D<float32_t> outImage;
+
+[numthreads(WORKGROUP_DIM, WORKGROUP_DIM, 1)]
```
Done.

Commit:
Use constant workgroup dimension instead of WORKGROUP_DIM define

Commit:
Remove passing WORKGROUP_DIM to shader

kevyuu commented on include/nbl/builtin/hlsl/sampling/warps/spherical.hlsl:
```diff
@@ -0,0 +1,80 @@
+		const density_type sinTheta = sqrt(density_type(1) - (cosTheta * cosTheta));
+		return density_type(1) / (sinTheta * density_type(2) * numbers::pi<density_type> * numbers::pi<density_type>);
+	}
+};
```
This is a warp. BijectiveSampler needs forwardPDF and backwardPDF method. Does it make sense for this class? Should I just return 1 / domainArea?

devshgraphicsprogramming replied:
Yeah it would basically be similar to the UniformHemisphere sampler, but oviously with a very different coordinate mapping

devshgraphicsprogramming replied:
It would have to be called something like `SphericalMappingSampler`

kevyuu commented on include/nbl/builtin/hlsl/sampling/hierarchical_image/gen_warp.comp.hlsl:
```diff
@@ -0,0 +1,48 @@
+    using LuminanceSampler = LuminanceMapSampler<float32_t, LuminanceAccessor>;
+
+    LuminanceSampler luminanceSampler = 
+      LuminanceSampler::create(luminanceAccessor, uint32_t2(lumaMapWidth, lumaMapHeight), lumaMapWidth != lumaMapHeight, uint32_t2(lumaMapWidth, lumaMapHeight));
```
Done.

Commit:
Remove passing WORKGROUP_DIM for gen_luma

Commit:
Add upsscale parameter for EnvmapSampler

Commit:
Pass warpMap width and height to luminanceSampler. Check For Out of Bound access

Commit:
Small fixes to gen_warp shader

Commit:
gen_warp get image dimension from push constant instead of OpImageQuery. Assert oob access

Commit:
Rename HierarchicalImage to WarpmapSampler

Commit:
Add todo comment for cube map

kevyuu commented on src/nbl/CMakeLists.txt:
```diff
@@ -127,7 +127,9 @@ unset(NABLA_HEADERS_PUBLIC2 ${NBL_TMP_FULL_PATHS})
 set(NBL_CORE_SOURCES
 	core/alloc/refctd_memory_resource.cpp
 	core/hash/blake.cpp
+	core/sampling/EnvmapSampler.cpp
```
Done.

Commit:
Move EnvmapSampler from core/sampling to video/sampling

Commit:
Rename LuminanceSampler to HierarchicalLuminanceSampler

kevyuu commented on include/nbl/builtin/hlsl/sampling/hierarchical_image.hlsl:
```diff
@@ -0,0 +1,193 @@
+		is_scalar_v<ScalarT> && 
+		hierarchical_image::LuminanceReadAccessor<LuminanceAccessorT, ScalarT>
+	)
+struct LuminanceMapSampler
```
Done.

Commit:
Fix corner sampling logic in gen_luma and hierarchical_image.hlsl

Commit:
Fix previous commit

Commit:
Add some todo comment for corner sampling flag

Commit:
Add const modifier for binarySearch

Commit:
Add some temporary struct from pr #1001

kevyuu commented on include/nbl/builtin/hlsl/sampling/hierarchical_image.hlsl:
```diff
@@ -0,0 +1,193 @@
+
+
+		// If we don`t add xi, the sample will clump to the lowest corner of environment map texel. We add xi to simulate uniform distribution within a pixel and make the sample continuous. This is why we compute the pdf not from the normalized luminance of the texel, instead from the reciprocal of the Jacobian.
+		const vector2_type directionUV = (vector2_type(p.x, p.y) + xi) / vector2_type(_mapSize);
```
Done.

kevyuu commented on include/nbl/builtin/hlsl/sampling/hierarchical_image.hlsl:
```diff
@@ -0,0 +1,193 @@
+				}
+      }
+      if (choseSecond(wx_0, wx_1, xi.x))
+        p.x |= 1;
```
Done.

Commit:
Refactor hierarchical_image naming and concepts

Commit:
Fix indentation of hierarchical_image.hlsl

Commit:
Small fixes

Commit:
Remove superfluous member from WarpmapSampler

Commit:
Merge branch 'master' into env_map_importance_sampling

Commit:
Fix resolve.hlsl case

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/hierarchical_image/gen_warp.comp.hlsl:
```diff
@@ -0,0 +1,46 @@
+[[vk::binding(0, 0)]] Texture2D<float32_t> lumaMap;
+
+[[vk::binding(1, 0)]] RWTexture2D<float32_t2> outImage;
+
```
arrays

devshgraphicsprogramming commented on include/nbl/video/sampling/EnvmapSampler.h:
```diff
@@ -0,0 +1,147 @@
+		{
+				core::smart_refctd_ptr<asset::IAssetManager> assetManager = nullptr;
+				core::smart_refctd_ptr<video::IGPUImageView> envMap = nullptr;
+				uint8_t upscaleLog2 = 0;
```
why are we baking the source image and warpmap generation arguments into the class that creates 2 pipelines?

Shouldn't the same shader be able to update warpmaps for multiple envmaps?

devshgraphicsprogramming commented on include/nbl/video/sampling/EnvmapSampler.h:
```diff
@@ -0,0 +1,147 @@
+			hlsl::uint32_t2 lumaWorkgroupCount;
+			hlsl::uint32_t2 warpWorkgroupCount;
+			core::smart_refctd_ptr<video::IGPUImageView> lumaMap;
+			core::smart_refctd_ptr<video::IGPUImageView> warpMap;
```
things to keep in your `SSession`

devshgraphicsprogramming commented on include/nbl/video/sampling/EnvmapSampler.h:
```diff
@@ -0,0 +1,147 @@
+		nbl::video::IGPUCommandBuffer::SPipelineBarrierDependencyInfo::image_barrier_t getLumaMapBarrier(
+			core::bitflag<nbl::asset::PIPELINE_STAGE_FLAGS> dstStageMask,
+			core::bitflag<nbl::asset::ACCESS_FLAGS> dstAccessMask,
+			nbl::video::IGPUImage::LAYOUT oldLayout);
```
All 3 will likely be accessed by same stages, except that Luma and Warpmap will only be read by all other shaders except for us (so you can strip the write access masks from provided ones)

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/hierarchical_image.hlsl:
```diff
@@ -0,0 +1,312 @@
+template <typename ScalarT, typename LuminanceAccessorT, typename PostWarpT 
+  NBL_PRIMARY_REQUIRES(
+    is_scalar_v<ScalarT> && 
+    hierarchical_image::LuminanceReadAccessor<LuminanceAccessorT, ScalarT> &&
```
LuminanceAccessorT should tell you the `PostWarpT` since the luma map had to be made with luminance weighed by `1/forwardPdf` of the post-warp, so the warp mustl be strictly paired with the  luma map

and then in turn the `PostWarpT` also tells you your `scalar_type` (its own typedef)

So only one tempalte parameter needed

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/hierarchical_image.hlsl:
```diff
@@ -0,0 +1,312 @@
+  using vector3_type = vector<ScalarT, 3>;
+  using vector4_type = vector<ScalarT, 4>;
+  using domain_type = vector2_type;
+  using codomain_type = vector3_type;
```
same as with the `ComposedHierarchicalSampler`

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/hierarchical_image/gen_luma.comp.hlsl:
```diff
@@ -0,0 +1,32 @@
+
+		const float uv_y = (float(threadID.y) + float(0.5f)) / pc.lumaMapHeight;
+		const float32_t3 envMapSample = envMap.Load(float32_t3(threadID.xy, 0));
+		float32_t luma = hlsl::dot(envMapSample, pc.lumaRGBCoefficients) * sin(numbers::pi<float32_t> * uv_y);
```
btw the `uv` going into the spherical warp is wrong (not accounting for corner sampling), but its a differential element anyway so its going to be approximate/inaccurate anyway (you'd want to use a full rectangle or triangle solid angle integral before worrying about the exact uv to a precision of half a texel)

put this in the comment

devshgraphicsprogramming replied:
still not done, aside from the luma clamping, but check if the constant is not too small because pdf = min/(min+min) may produce a subnormal number

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/hierarchical_image.hlsl:
```diff
@@ -0,0 +1,312 @@
+  LuminanceAccessorT _map;
+  float32_t _rcpAvgLuma;
+  float32_t2 _rcpWarpSize;
+  uint16_t2 _mapSize;
```
you shold really store `_lastTexel` (`_mapSize-1`) instead of `mapSize`

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/hierarchical_image.hlsl:
```diff
@@ -0,0 +1,193 @@
+		const vector2_type texelCoord = xi * float32_t2(_lastWarpPixel);
+
+		matrix<scalar_type, 4, 2> uvs = _warpMap.sampleUvs(uint32_t2(texelCoord));
+
```
my bad it totally wont work, needs to take unnormalized but floating point after corner sampling is applied

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/hierarchical_image.hlsl:
```diff
@@ -0,0 +1,312 @@
+      yDiff // second column dFdy
+    )) * _effectiveWarpArea;
+
+    const scalar_type pdf = abs(warpResult.density / detInterpolJacobian);
```
`postWarp` density, `xDiffs`, `yDiff` and `interpolant.y` go in the `cache` between `generate` and `forwardPdf`

devshgraphicsprogramming replied:
also the `uv` put into the post warp (for the forward weight function)

Commit:
Merge branch 'master' into env_map_importance_sampling

kevyuu committed:
Rename GenWarpWorkgroupDim and GenLumaWorkgroupDim

kevyuu committed:
Fix include guard of warps/spherical.hlsl

kevyuu committed:
Undo Embed built-in resources option to OFF

kevyuu committed:
Gen Luma now support Texture Array

kevyuu committed:
Remove unused member in HierarchicalWarpGenerator

kevyuu committed:
Remove LuminanceReadAccessor concept and support Texture2DArray for gen_warpmap

kevyuu committed:
Rename _mip2x1 to _lastMipLevel

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/hierarchical_image.hlsl:
```diff
@@ -75,20 +76,20 @@ struct HierarchicalWarpGenerator
-  float32_t2 _rcpWarpSize;
   uint16_t2 _mapSize;
+  uint16_t _layerIndex;
   uint16_t _mip2x1 : 15;
```
max mip level is 15, you only need 4 bits to store it, leaving you  with 11 bits for the layer, and it just so happens that max layer count is 2048 (where the max value needs only 11 bits) on most GPUs

all 3 members fit in one uint16

kevyuu committed:
Fix envmapsampler header guard

kevyuu committed:
Remove unused defaults in EnvmapSampler

kevyuu committed:
Undo cuda interop cmake option

kevyuu committed:
Improve comment on hierarchical_image

kevyuu committed:
Using sincos in spherical.hlsl

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/hierarchical_image.hlsl:
```diff
@@ -75,20 +76,20 @@ struct HierarchicalWarpGenerator
   uint16_t2 _mapSize;
-  uint16_t _mip2x1 : 15;
+  uint16_t _layerIndex;
+  uint16_t _lastMipLevel : 15;
```
its not the last, its the penultimate (one minus last) because you're not after the last one which is 1x1, you're tapping 2x1 or 2x2

kevyuu committed:
Rework hierarchical_image and sphericalSampling to follow new sampler concept.

kevyuu commented on include/nbl/builtin/hlsl/sampling/hierarchical_image.hlsl:
```diff
@@ -0,0 +1,312 @@
+
+  V _value;
+  P _pdf;
+};
```
Done. Removed.

kevyuu committed:
refactor: Improve `generate` function and sampling logic

The `generate` function in `HierarchicalWarpGenerator` was updated:
- Replaced `xi` parameter with `v` and introduced a local `xi`.
- Adjusted `directionUV` calculation to normalize coordinates by
  subtracting `0.5` and dividing by `_lastTexel` for improved accuracy.
- Added a comment to clarify the normalization process.

Commit by Kevin Yudi Utama

kevyuu committed:
refactor: Update backwardPdf and backwardWeight logic

Commit by Kevin Yudi Utama

kevyuu committed:
refactor: Improve hierarchical image sampling system

Refine concepts and simplify calculations in hierarchical image sampling.

kevyuu committed:
refactor: Implement MipmappedLuminanceReadAccessor

- Remove `ScalarT` template parameter and derive `scalar_type` from `LuminanceAccessorT`.
- Remove `_layerIndex` from `HierarchicalWarpGenerator` and encapsulate layer handling in `LuminanceAccessorT`.
- Use `getAvgLuma` from `LuminanceAccessorT` instead of external `rcpAvgLuma`.
- Introduce `MipmappedLuminanceReadAccessor` concept to define accessor requirements.
- Refactor `LuminanceAccessor` to encapsulate layer index and provide methods like `get`, `resolution`, and `getAvgLuma`.
- Update `gen_warp.comp.hlsl` to use the refactored `LuminanceAccessor` and `HierarchicalWarpGenerator` interfaces.

kevyuu committed:
refactor: Embed value_type inside LuminanceReadAccessor

kevyuu committed:
refactor: embed scalar_type inside WarpAccessor

kevyuu committed:
fix: Forward Weight calculation in WarpMapSampler

kevyuu committed:
fix: Implement bacwardWeight for WarpGenerator

kevyuu committed:
refactor: Rename HierarchicalWarpGenerator to HierarchicalLuminanceSampler

kevyuu committed:
Rename HierarchicalWarpSampler to ComposedHierarchicalSampler

kevyuu commented on include/nbl/builtin/hlsl/sampling/hierarchical_image.hlsl:
```diff
@@ -0,0 +1,312 @@
+    hierarchical_image::LuminanceReadAccessor<LuminanceAccessorT, ScalarT> &&
+    concepts::Warp<PostWarpT>
+  )
+struct HierarchicalWarpSampler
```
Done.

kevyuu committed:
Remove deleted hlsl from list of builtin resource in CMake

kevyuu committed:
Add constraints and improve type consistency for HierarchicalLuminanceSampler

kevyuu committed:
Move WarpmapSampler into its own file

kevyuu committed:
Fix include guard of hierarchical_image/accesors.hlsl

kevyuu committed:
Unify pipelineLayout and descriptorSet for genLuma and genWarp

kevyuu committed:
Remove mount method in EnvmapSampler

No longer extension, so no need to mount.

kevyuu committed:
Implement common pipeline creation function

kevyuu committed:
Added SSession on EnvmapSampler

The purpose of SSession is so that we don't need to recreate new pipeline from scratch for each envmap.;

kevyuu committed:
Rename EnvmapSampler CEnvmapWarpGenerator

kevyuu committed:
Move LumaRgBCoeffecients as constant in hlsl header

kevyuu committed:
Implement multi layer gen_luma and gen_warp

kevyuu committed:
Remove non write access from srcAccessMask in image barriers

kevyuu committed:
Update submodule to latest master

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/spherical_mapping.hlsl:
```diff
@@ -0,0 +1,92 @@
+
+  weight_type forwardWeight(const domain_type v, const cache_type cache)
+  {
+    return scalar_type(1) / (scalar_type(2) * cache.sinTheta * numbers::pi<scalar_type> *numbers::pi<scalar_type>);
+  }
+
+  density_type backwardPdf(const codomain_type v)
+  {
+    const density_type cosTheta = v.y;
+    const density_type rcpSinTheta = hlsl::rsqrt(density_type(1) - (cosTheta * cosTheta));
+    return rcpSinTheta / (density_type(2) * numbers::pi<density_type> * numbers::pi<density_type>);
+  }
+
+  weight_type backwardWeight(const codomain_type v)
+  {
+    const density_type cosTheta = v.y;
+    const density_type rcpSinTheta = hlsl::rsqrt(density_type(1) - (cosTheta * cosTheta));
+    return rcpSinTheta / (density_type(2) * numbers::pi<density_type> * numbers::pi<density_type>);
+  }
```
make the weight call the PDF function so there's no duplicate code

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/hierarchical_image/common.hlsl:
```diff
@@ -0,0 +1,42 @@
+  uint32_t warpMapWidth : 16;
+  uint32_t warpMapHeight : 16;
+  // Both warpMap and lumaMap should have the same layer count
+  uint16_t lumaMapLayer;
```
you cannot use uint16 in push constants (AMD), have to use uint32_t and use a bitfield instead

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/hierarchical_image/common.hlsl:
```diff
@@ -0,0 +1,42 @@
+namespace hierarchical_image
+{
+
+NBL_CONSTEXPR_INLINE_NSPC_SCOPE_VAR float32_t3 LumaRgbCoefficients = { 0.2126729f, 0.7151522f, 0.0721750f };
```
use numbers from the OETF/EOTF headers

devshgraphicsprogramming commented on include/nbl/builtin/hlsl/sampling/warpmap.hlsl:
```diff
@@ -0,0 +1,125 @@
+{
+
+// TODO: Add some constraint into PostWarpT
+// Ask(kevin): Should we add constraint so the WarpAccessor::scalar_type is the same as LuminanceAccessorT::value_type. One is a uv and the other is luminance. Technically, they can have different type. 
```
no we shouldn't, luma could be float32_t due to extreme HDR but warpmap could be unorm16 / float16_t if the luma is low res enough

kevyuu committed:
Add some comment regarding choseSecond implementation

kevyuu committed:
Fix bug in hierarchical_image.hlsl

kevyuu committed:
Fix bug in gen_warp where we reverse the height and width

kevyuu committed:
Use luma dimension when caclulate the mip level

kevyuu committed:
Remove unecessary comments

kevyuu committed:
Add correction uv logic when fetching luma when calculating for mis weights

kevyuu committed:
Add some debug marker in CEnvmapWarpGenerator.cpp

kevyuu committed:
Fix bug when blitting image with IGPUImageView::remaining_array_layers

kevyuu committed:
Fix indentation of CEnvmapWarpGenerator.cpp

kevyuu committed:
Change push constant member to uint32_t

kevyuu committed:
Merge branch 'master' into env_map_importance_sampling

kevyuu committed:
Make weight calc method call pdf calc method

kevyuu committed:
Resolve CMakeLists.txt merge conflict

kevyuu committed:
Refactor luma calculation to use Colorspace module

kevyuu committed:
Remove outdated comment

kevyuu committed:
Add BackwardTractableSampler concept PostWarpT

kevyuu committed:
Constraint PostWarpT to BijectiveSampler

kevyuu committed:
Add some ask comment

kevyuu committed:
Use PostWarpT in gen_luma
